### ISSUE TITLE
After http training, server throws these errors until restart

### ISSUE THREAD

**Issue by atotalnoob** [RasaHQ/rasa_nlu#1239]
**Rasa NLU version**: 0.12.2

**Operating system** (windows, osx, ...): Windows Server 2012 R2

**Content of model configuration file**:
Server YML:
language: "en"

pipeline:
- name: "nlp_spacy"
- name: "tokenizer_spacy"
- name: "intent_featurizer_spacy"
- name: "ner_crf"
- features: [["low", "title"], ["bias", "word3"], ["upper", "pos", "pos2"]]
- name: "ner_synonyms"
- name: "intent_classifier_sklearn"
- name: "intent_entity_featurizer_regex"
- name: "ner_duckling"
  dimensions: [ "time", "number", "duration"] 

Training YML: 
language: "en"

pipeline:
- name: "nlp_spacy"
- name: "tokenizer_spacy"
- name: "intent_featurizer_spacy"
- name: "ner_crf"
  features: [["low", "title"], ["bias", "word3"], ["upper", "pos", "pos2"]]
- name: "ner_synonyms"
- name: "intent_classifier_sklearn"
- name: "intent_entity_featurizer_regex"
- name: "ner_duckling"
  dimensions: [ "time", "number", "duration"] 

data: {Data Goes Here}

**Issue**: After a remote training request is completed (new model trained) this error is thrown, the server stops responding to anything, status, parsing, etc.

Anonymized Stacktrace:
```
2018-07-16 11:36:19-0400 [-] 2018-07-16 11:36:19 ERROR    __main__ - java.lang.N
ullPointerException
2018-07-16 11:36:19-0400 [-] Traceback (most recent call last):
2018-07-16 11:36:19-0400 [-]   File "C:\chatbot\App_Data\rasa\rasa_nlu\server.py
", line 240, in parse
2018-07-16 11:36:19-0400 [-]     else threads.deferToThread(self.data_router.par
se, data))
2018-07-16 11:36:19-0400 [-]   File "C:\ProgramData\Anaconda3\lib\site-packages\
twisted\python\threadpool.py", line 250, in inContext
2018-07-16 11:36:19-0400 [-]     result = inContext.theWork()
2018-07-16 11:36:19-0400 [-]   File "C:\ProgramData\Anaconda3\lib\site-packages\
twisted\python\threadpool.py", line 266, in <lambda>
2018-07-16 11:36:19-0400 [-]     inContext.theWork = lambda: context.call(ctx, f
unc, *args, **kw)
2018-07-16 11:36:19-0400 [-]   File "C:\ProgramData\Anaconda3\lib\site-packages\
twisted\python\context.py", line 122, in callWithContext
2018-07-16 11:36:19-0400 [-]     return self.currentContext().callWithContext(ct
x, func, *args, **kw)
2018-07-16 11:36:19-0400 [-]   File "C:\ProgramData\Anaconda3\lib\site-packages\
twisted\python\context.py", line 85, in callWithContext
2018-07-16 11:36:19-0400 [-]     return func(*args,**kw)
2018-07-16 11:36:19-0400 [-]   File "C:\chatbot\App_Data\rasa\rasa_nlu\data_rout
er.py", line 242, in parse
2018-07-16 11:36:19-0400 [-]     model)
2018-07-16 11:36:19-0400 [-]   File "C:\chatbot\App_Data\rasa\rasa_nlu\project.p
y", line 118, in parse
2018-07-16 11:36:19-0400 [-]     interpreter = self._interpreter_for_model(model
_name)
2018-07-16 11:36:19-0400 [-]   File "C:\chatbot\App_Data\rasa\rasa_nlu\project.p
y", line 196, in _interpreter_for_model
2018-07-16 11:36:19-0400 [-]     return Interpreter.create(metadata, self._compo
nent_builder)
2018-07-16 11:36:19-0400 [-]   File "C:\chatbot\App_Data\rasa\rasa_nlu\model.py"
, line 303, in create
2018-07-16 11:36:19-0400 [-]     model_metadata, **context)
2018-07-16 11:36:19-0400 [-]   File "C:\chatbot\App_Data\rasa\rasa_nlu\component
s.py", line 398, in load_component
2018-07-16 11:36:19-0400 [-]     cached_component, **context)
2018-07-16 11:36:19-0400 [-]   File "C:\chatbot\App_Data\rasa\rasa_nlu\registry.
py", line 132, in load_component_by_name
2018-07-16 11:36:19-0400 [-]     return component_clz.load(model_dir, metadata,
cached_component, **kwargs)
2018-07-16 11:36:19-0400 [-]   File "C:\chatbot\App_Data\rasa\rasa_nlu\extractor
s\duckling_extractor.py", line 194, in load
2018-07-16 11:36:19-0400 [-]     duckling = cls.create_duckling_wrapper(language
)
2018-07-16 11:36:19-0400 [-]   File "C:\chatbot\App_Data\rasa\rasa_nlu\extractor
s\duckling_extractor.py", line 108, in create_duckling_wrapper
2018-07-16 11:36:19-0400 [-]     return DucklingWrapper(language=language)
2018-07-16 11:36:19-0400 [-]   File "C:\ProgramData\Anaconda3\lib\site-packages\
duckling\wrapper.py", line 36, in __init__
2018-07-16 11:36:19-0400 [-]     self.duckling.load()
2018-07-16 11:36:19-0400 [-]   File "C:\ProgramData\Anaconda3\lib\site-packages\
duckling\duckling.py", line 105, in load
2018-07-16 11:36:19-0400 [-]     duckling_load.invoke()
2018-07-16 11:36:19-0400 [-] jpype._jexception.java.lang.NullPointerExceptionPyR
aisable: java.lang.NullPointerException
2018-07-16 11:36:19-0400 [-] "10.106.28.153" - - [16/Jul/2018:15:36:19 +0000] "G
ET /parse?q=ParseText&project=projectName HT
TP/1.1" 500 47 "-" "RestSharp/106.3.1.0"

```

**Comment by akelad** [comment]
Hmm, seems to be some problem with duckling. Can you try using `ner_duckling_http` instead?

**Comment by atotalnoob** [comment]
This solved the issue. 

I updated ner_duckling to ner_duckling_http, added the url and kept the same dimension.

**Comment by atotalnoob** [comment]
*bump*

**Comment by akelad** [comment]
But you're using remote training, why can't you use a remote duckling then? To use local duckling you have to use `ner_duckling`, but I'm not sure what's going on with your version of duckling there. I guess try reinstalling it?

**Comment by atotalnoob** [comment]
Remote training stays within my company's network. If I use duckling http, it leaves my company's network (also its blocked by the firewall). My company would NOT be okay with their information leaving the network.

I have already tried reinstalling duckling. I have duckling version 1.8.0. 

What else can I try? I can set up a remote debugging session..

**Comment by atotalnoob** [comment]
Definitely something wrong with Rasa NLU, I tested Duckling on its own and it works fine.

**Comment by atotalnoob** [comment]
Okay, I've narrowed it down. 

It only happens when a server is launched with pre_load all, then a new model is trained, and the new model is attempted to use. So launching the server without pre_load all everything works fine if the model is not loaded, then you can train as many models as your heart desires and the issue does not reoccur (even if you load a model by asking it to parse and then train a new one). 

This is the block that is throwing an error:
```
 if languages:
            # Duckling's load function expects ISO 639-1 Language Codes (e.g. "en")
            iso_languages = [Language.convert_to_iso(lang) for lang in languages]
            duckling_load.invoke(
                clojure_hashmap.invoke(
                    self.clojure.read(':languages'),
                    clojure_list.invoke(iso_languages)
                )
            )
        else:
            duckling_load.invoke()

```
I've tried bypassing the if block and running `duckling_load.invoke()` directly. Same issue.

**Comment by atotalnoob** [comment]
Trying to attach a debugger to server.py, but everytime I do, I get the an attributeError: 'datarouter' object has no attribute '_pre_load'

**Comment by akelad** [comment]
I think you talked about this with someone on Gitter as well? And found out you can run duckling http, because it doesn't actually leave the network. How's that going?

**Comment by atotalnoob** [comment]
That isn't fixing a problem, its a workaround. The issue still remains.

**Comment by atotalnoob** [comment]
Bump

**Comment by akelad** [comment]
Well in fairness we do in general suggest using the http version because it actually works better. But i'm really not sure about what that error is, maybe @tmbo has an idea?

**Comment by atotalnoob** [comment]
Not trying to be rude, but your docs don't mention duckling http at all....

**Comment by tmbo** [comment]
The connection between the JVM (used by the `ner_duckling` component) and python is quite brittle. The recommended way really is to use `ner_duckling_http` (will be part of the updated docs we will release in a couple of days) - the duckling server doesn't necessarily need to be remote. Depending on your setup (non of these will move any information over the wire):

1. you can install duckling within your docker container if you use docker, and use `localhost:8000` as the remote url
2. use docker compose to combine a rasa nlu docker container with a prebuild duckling docker container 
3. install `stack` and `duckling` locally on the machine (if you installed rasa nlu directly without docker)

**Comment by wrathagom** [comment]
Lot's of progress on various areas of this thread:

The [0.13 release notes](https://stackoverflow.com/a/51738593/1408476) mention that support for `ner_duckling` will be completely dropped in the next version. And the [docs have been updated](http://www.rasa.com/docs/nlu/pipeline/#ner-duckling-http) to mentioned `ner_duckling_http` and there is a[ compose example](https://github.com/RasaHQ/rasa_nlu/blob/master/docker/docker-compose.yml) for bringing up duckling in docker with Rasa.

@atotalnoob let us know if you are still having issues with duckling. Or create a new issue for more assistance.
